activerecord/lib/active_record/connection_adapters/schema_cache.rb

```diff
@@ -129,10 +129,29 @@ def marshal_dump
 
       def marshal_load(array)
         @version, @columns, @columns_hash, @primary_keys, @data_sources, @indexes, @database_version = array
-        @indexes = @indexes || {}
+        @indexes ||= {}
+
+        @columns = deep_deduplicate(@columns)
+        @columns_hash = deep_deduplicate(@columns_hash)
```

Comment:
@kaspth another thing: I think we should simply stop storing `columns_hash` in the schema_cache dump, it's very easily rebuilt from `columns`. 

That would make parsing much faster, and would save lots of allocations.

Reply:
I’d definitely be game to see a PR for that!